src/08-recommendations/03-pipeline-recommend.ts

```diff
@@ -0,0 +1,325 @@
+  const query = `
+    query ArtworkRecommendations($userID: String!, $size: Int!, $includeBackfill: Boolean!) {
+      who: me {
+        email
+      }
+      artworksForUser(userId: $userID, includeBackfill: $includeBackfill, first: $size) {
+        edges {
+          node {
+            ...artwork
+          }
+        }
+      }
+      me {
+        artworkRecommendations(first: $size) {
+          edges {
+            node {
+              ...artwork
+            }
+          }
+        }
+        recommendedArtworks(first: $size) {
+          edges {
+            node {
+              ...artwork
+            }
+          }
+        }
+      }
+    }
+
+    fragment artwork on Artwork {
+      slug
+      title
+      date
+      artist {
+        name
+        gender
+        nationality
+        birthday
+      }
+      rarity: attributionClass {
+        name
+      }
+      medium: mediumType {
+        name
+      }
+      materials: medium
+      price: saleMessage
+      # dominantColors
+      # href
+    }
+  `
+
+  const variables = {
+    userID: userID,
+    size: NUM_CANDIDATES_PER_FIELD,
+    includeBackfill: INCLUDE_BACKFILL,
+  }
+
+  const headers = {
+    "X-Access-Token": accessToken!,
+  }
+
+  const response = await metaphysics({ query, variables, headers })
+
+  // hard to know which recommendation fields will be populated,
+  // so for now just fetch them all and concatenate them
+  const artworksToUse = [
+    ...response.data.me.recommendedArtworks.edges,
+    ...response.data.me.artworkRecommendations.edges,
+    ...response.data.artworksForUser.edges,
+  ]
+
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  candidateArtworks = artworksToUse.map(({ node }: any) => node)
+
+  return candidateArtworks
+}
+
+const tools: OpenAI.Chat.ChatCompletionTool[] = [
+  {
+    type: "function",
+    function: {
+      name: "getArtworkRecommendations",
+      description:
+        "Get a list of artworks that Artsy has recommended for the user (based on existing recommendation pipelines)",
+      parameters: {
+        type: "object",
+        properties: {},
+      },
+    },
+  },
+]
+
+async function metaphysics(args: {
+  query: string
+  variables: Record<string, unknown>
+  headers: Record<string, string>
+}) {
+  const { query, variables, headers } = args
+
+  console.log(chalk.black("\nQuerying Metaphysics"))
+  VERBOSE && console.log(chalk.black.dim(JSON.stringify({ variables })))
+  VERBOSE && console.log(chalk.black.dim(query))
+
+  const url = "https://metaphysics-staging.artsy.net/v2"
+  const body = JSON.stringify({ query, variables })
+  const options = {
+    method: "POST",
+    headers: { ...headers, "Content-Type": "application/json" },
+    body,
+  }
+
+  const response = await fetch(url, options)
+  const json = await response.json()
+  VERBOSE && console.log(JSON.stringify(json, null, 2))
+  return json
+}
+
+function printStatistics(content: string | null) {
```

Comment:
Found it helpful to assess positional bias with a few simple stats logged out — see bottom of screencap in the PR description.

(Was more of an issue with the first script than this one, but I've kept it here, since I never committed this into the first one)

Reply:
Like it 🙌